### PR TITLE
specification: Security model

### PR DESCRIPTION
**Pull request by sameo**

### CODE REVIEW

**Comment by sameo** [comment]
cc @rsahita

File: specification/security_model.adoc [hunk 1]
```diff
@@ -1,9 +1,195 @@
...
+- Only the TVM owner can assess of a TDI trustworthiness. Based on that
+  assessment, it explicitly accepts or rejects a TDI into its TCB.
+- A TDI may access a TVM confidential memory only when the following conditions
+  are met:
+  1. The TVM owner has explicitly allowed the TDI to access its memory by
+     accepting it.
+  2. The TDI is exclusively assigned to the TVM, i.e. it must not be shared
+     other TVMs or any host software component.
+- A TVM may trust and access a TDI MMIO space if and only if the TDI is
+  exclusively assigned to the TVM.
```
Comment (jyao1) [inline specification/security_model.adoc:15]: I feel this sentence need reword.

The original sentence implies that "as long as TDI is exclusively assigned, then the TVM may trust it." I don't think that is true, because the TVM need accept the TDI as well as prerequisite.

I think DMA and MMIO should have similar rule:
1. The TVM owner has explicitly accepted the TDI.
2. The TDI is exclusively assigned to the TVM.

File: specification/security_model.adoc [hunk 2]
```diff
@@ -1,9 +1,195 @@
...
+
+- Only the TVM owner can assess of a TDI trustworthiness. Based on that
+  assessment, it explicitly accepts or rejects a TDI into its TCB.
+- A TDI may access a TVM confidential memory only when the following conditions
+  are met:
+  1. The TVM owner has explicitly allowed the TDI to access its memory by
+     accepting it.
+  2. The TDI is exclusively assigned to the TVM, i.e. it must not be shared
+     other TVMs or any host software component.
+- A TVM may trust and access a TDI MMIO space if and only if the TDI is
```
Comment (jyao1) [inline specification/security_model.adoc:14]: Here "TDI MMIO" means "trusted MMIO", right?
A device may have "untrusted MMIO", which does not follow this rule.

File: specification/security_model.adoc [hunk 3]
```diff
@@ -1,9 +1,195 @@
...
-== Security and Threat Model
+== Security Model
+
+The CoVE-IO securiy model is built on the following assumptions:
+
+- Only the TVM owner can assess of a TDI trustworthiness. Based on that
+  assessment, it explicitly accepts or rejects a TDI into its TCB.
+- A TDI may access a TVM confidential memory only when the following conditions
+  are met:
+  1. The TVM owner has explicitly allowed the TDI to access its memory by
```
Comment (jyao1) [inline specification/security_model.adoc:10]: Here "its memory" means "TVM private memory", right?
A TVM may have "shared memory", which does not follow this rule.

File: specification/security_model.adoc [hunk 4]
```diff
@@ -1,9 +1,195 @@
...
+based on TVM specific set of verification policies and criteria. A TVM accepting
+a TDI does not imply that other TVMs on the same host platform would accept it
+as well.
+
+Multi-function, CoVE-IO-compliant devices may expose multiple TDIs assigned to
+different TVMs. It is the DSM responsibility to guarantee isolation between all
+assigned TDIs, on a per-TVM basis.
+
+TDIs must not be shared between TVMs, they must be exclusively assign to zero or
+at most one TVM.
```
Comment (jyao1) [inline specification/security_model.adoc:28]: I am not sure that using "TDIs" is right way. Maybe using "any TDI"?

E.g. "Any TDI must not be shared between TVMs, it must be exclusively ..."

Comment (jyao1) [reply specification/security_model.adoc:28]: Maybe we can clarify that it is legal to assign multiple TDIs to one TVM.

File: specification/security_model.adoc [hunk 5]
```diff
@@ -1,9 +1,195 @@
...
+single selective IDE stream between the host platform and any physical device
+from which one or more TDIs are assigned to TVMs. All TDIs withing that physical
+device share the single PCIe IDE stream.
+
+The main security objective of the CoVE-IO security model is to protect a TVM's
+confidential data integrity and confidentiality while TDIs are assigned to it.
+At the same time, availability of those assigned TDIs is out of this model scope
+as e.g. the host VMM could remove them from the TVM at any time it sees fit. All
+TVM confidential data must be cleared and wiped before the host software stack
+can fully reclaim an assigned TDI.
```
Comment (jyao1) [inline specification/security_model.adoc:42]: I think you mean "all TVM confidential data **in the TDI** must be cleared...".
There is no need to clear the TVM confidential data in the TVM, to reassign a TDI.

File: specification/security_model.adoc [hunk 6]
```diff
@@ -1,9 +1,195 @@
...
+
+=== Assets
+
+The CoVE-IO security model aims at protecting the following assets:
+
+1. The TVM confidential data, which includes its confidential main memory, code
+   and state.
+2. Any assigned device I/O space that is mapped to the TVM address space,
+   also known as device MMIO.
+3. System memory.
```
Comment (jyao1) [inline specification/security_model.adoc:54]: why "3. system memory" ?

File: specification/security_model.adoc [hunk 7]
```diff
@@ -1,9 +1,195 @@
...
+
+* _Privileged host software adversary_: This includes host software components
+  executing in S and M mode like the host firmware, kernel, hypervisor, VMM,
+  etc. As the system resource owner but also the TVMs lifecycle manager, those
+  components can access and control all devices on the system.
+
+* _Unprivileged host software adversary_: (HS mode??)
+
+* _Privileged guest software adversary_: (Other TVM guest kernels)
+
```
Comment (jyao1) [inline specification/security_model.adoc:69]: why we miss "Unprivileged guest software adversary" ?

File: specification/security_model.adoc [hunk 8]
```diff
@@ -1,9 +1,195 @@
...
+==== CoVE-IO-T003
+
+.CoVE-IO-T003
+|===
+| Asset | Threat | Adversary | Scope | Result
+
+| TVM confidential data
+| Denial of service
+| Privileged host software
+| **Not** in scope
```
Comment (jyao1) [inline specification/security_model.adoc:127]: I am not sure if we really need list this.
We said before that availability is our of scope.

File: specification/security_model.adoc [hunk 9]
```diff
@@ -1,9 +1,195 @@
...
+
+|===
+
+==== CoVE-IO-T004
+
+.CoVE-IO-T004
+|===
+| Asset | Threat | Adversary | Scope | Result
+
+| System memory
```
Comment (jyao1) [inline specification/security_model.adoc:142]: I am not sure if we want to list it here.
It seems irrelevant to confidential computing.

File: specification/security_model.adoc [hunk 10]
```diff
@@ -1,9 +1,195 @@
...
+
+|===
+
+==== CoVE-IO-T006
+
+.CoVE-IO-T006
+|===
+| Asset | Threat | Adversary | Scope | Result
+
+| Device MMIO
```
Comment (jyao1) [inline specification/security_model.adoc:180]: It should be "device trusted-MMIO", right?

A shared-MMIO does not have security properties.

File: specification/security_model.adoc (hunk 2 above)
Comment (sameo) [reply specification/security_model.adoc:14]: Yes, but how would you define untrusted MMIO? A part of the TDI MMIO space that has not been mapped through the TSM controlled (and confidential memory hold) second page table (equivalent to secure EPT) ?

Comment (jyao1) [reply specification/security_model.adoc:14]: According to TDISP spec, the device may return DEVICE_INTERFACE_REPORT for each MMIO_RANGE with bit IS_NON_TEE_MEM. 
* IS_NON_TEE_MEM = 0, means the MMIO is trusted MMIO.
* IS_NON_TEE_MEM = 1, means the MMIO is untrusted MMIO.

File: specification/security_model.adoc (hunk 6 above)
Comment (sameo) [reply specification/security_model.adoc:54]: I mean that `host memory that is not assigned to a TVM`, as a TVM with assigned TDIs could potentially use the TDI to tamper with that part of host memory (See T004). Any advice on how we could define that asset?

File: specification/security_model.adoc (hunk 7 above)
Comment (sameo) [reply specification/security_model.adoc:69]: Because I am not sure how unprivileged guest software could threat the defined assets. i.e. I dont think it actually is an adversary.

File: specification/security_model.adoc (hunk 9 above)
Comment (sameo) [reply specification/security_model.adoc:142]: It is a valid threat for any directly assigned device, not only for confidential computing, I agree. I am calling it out here as this is a threat that is not covered by the CoVE threat model.

File: specification/security_model.adoc (hunk 2 above)
Comment (sameo) [reply specification/security_model.adoc:14]: Makes sense. Then we need to define that notion before using it.

File: specification/security_model.adoc (hunk 9 above)
Comment (jyao1) [reply specification/security_model.adoc:148]: I agree that this is something need to be considered in a full system design.
I suggest we don't mix them with threat in confidential computing.

The reason is that: If we want to add all threat for the full system, then we need list more. for example:
hypervisor need prevent attack from OS.
OS need prevent attack from APP.
hypervisor need prevent attack from BIOS.
OS need prevent attack from BIOS.
hypervisor need prevent attack from device DMA.
OS need prevent attack from device DMA.
BIOS need prevent attack from device DMA.
......

I am not sure what is criteria to list or not list them?

File: specification/security_model.adoc (hunk 6 above)
Comment (jyao1) [reply specification/security_model.adoc:60]: Even if the host memory is tampered, it won't impact the confidentiality or integrity of TVM.

I recommend we **only** focus on the asset for the confidential computing. Otherwise, we have too many things need to   consider and we may lose focus.

File: specification/security_model.adoc (hunk 2 above)
Comment (sameo) [reply specification/security_model.adoc]: This should be clearer now.

File: specification/security_model.adoc (hunk 3 above)
Comment (sameo) [reply specification/security_model.adoc]: Fixed, thanks.

File: specification/security_model.adoc (hunk 4 above)
Comment (sameo) [reply specification/security_model.adoc]: Makes sense, thanks. It should be fixed now.

File: specification/security_model.adoc (hunk 5 above)
Comment (sameo) [reply specification/security_model.adoc]: Good points, both addressed now.

File: specification/security_model.adoc (hunk 6 above)
Comment (sameo) [reply specification/security_model.adoc]: Yes, I agree now. I removed that asset, and I removed the related threat.

File: specification/security_model.adoc [hunk 11]
```diff
@@ -1,9 +1,195 @@
...
+
+== Threat Model
+
+=== Assets
+
+The CoVE-IO security model aims at protecting the following assets:
+
+1. The TVM confidential data, which includes its confidential main memory, code
+   and state.
+2. Any assigned device I/O space that is mapped to the TVM address space,
```
[the remark this answers is not in the document]
Comment (sameo) [reply specification/security_model.adoc]: Fixed as well.

File: specification/security_model.adoc (hunk 7 above)
Comment (jyao1) [reply specification/security_model.adoc:75]: OK

File: specification/security_model.adoc [hunk 12]
```diff
@@ -1,9 +1,242 @@
...
+
+==== CoVE-IO-T007
+
+.CoVE-IO-T007
+|===
+
+| Asset | Threat | Adversary | Scope | Result
+
+| TVM confidential data
+| Tamper and Disclosure
```
Comment (jyao1) [inline specification/security_model.adoc:210]: I notice that T001 and T002 separate Tamper and Disclosure. Here they are combined in T007. 

Question: What is the criteria to separate or combine?

File: specification/security_model.adoc [hunk 13]
```diff
@@ -1,9 +1,242 @@
...
+
+| Device trusted MMIO
+| Denial of service
+| Privileged host software
+| In scope
+| TVM can not access the trusted MMIO space of a TDI that is assigned to it
+
+5+| **Description** - A privileged host software component configures PCIe
+    switches to redirect or drop MMIO accesses from the TVM to one of its
+    assigned TDIs.
```
Comment (jyao1) [inline specification/security_model.adoc:177]: Is this really in scope?

The VMM can touch the device and make IDE into insecure state **any time**.

I am not sure how that can be mitigated...

Comment (sameo) [reply specification/security_model.adoc]: I changed that one to be about the host VMM redirecting MMIO access from a TVM to an untrusted device or to another, unassigned device. This can be mitigated with the mandatory usage of selective stream and the protection of the RP IDE stream registers.

File: specification/security_model.adoc (hunk 12 above)
Comment (sameo) [reply specification/security_model.adoc:215]: Good point. I am not sure what is the preferred way, but I think we should combine? T001 and T002 can be achieved through the same vector. What do you think?

Comment (jyao1) [reply specification/security_model.adoc:215]: Yes, I am OK to merge, if the attack mechanisms are very similar.

But if the attack mechanisms are different, then we should separate them.

File: specification/security_model.adoc (hunk 13 above)
Comment (jyao1) [reply specification/security_model.adoc]: Yes, I am OK to change the threat to "tamper".

I just feel uncomfortable to see "denial of service", because it is almost impossible to mitigate.

File: specification/security_model.adoc (hunk 12 above)
Comment (sameo) [reply specification/security_model.adoc:198]: I merged 001 and 002 together.

File: specification/security_model.adoc (hunk 7 above)
Comment (rsahita) [reply specification/security_model.adoc:77]: if priv guest sw is an adversary then unpriv guest sw is also an adversary (since it depends on the priv guest sw for its access-control) - i think we should list it and in the threat mitigations specify why it is is mitigated.

File: specification/security_model.adoc [hunk 14]
```diff
@@ -1,9 +1,316 @@
 [[security_model]]
-== Security and Threat Model
+== Security Model
+
+The CoVE-IO securiy model is built on the following assumptions:
+
+- Only the TVM owner can assess of a TDI trustworthiness. Based on that
+  assessment, it explicitly accepts or rejects a TDI into its TCB.
```
Comment (rsahita) [inline specification/security_model.adoc:7]: we may want to explicitly state that - 
Until a TDI is accepted by the TVM: a.) no direct confidential memory access is allowed by the TDI, and b.) trusted MMIO access to the TDI is blocked by TSM. Untrusted MMIO access to the TDI may still be allowed by the VMM for ordinary VMs.

File: specification/security_model.adoc [hunk 15]
```diff
@@ -1,9 +1,321 @@
...
+  etc. As the system resource owner but also the TVMs lifecycle manager, those
+  components can access and control all devices on the system.
+
+* _Unprivileged host software adversary_: This includes host software components
+  executing in U mode like e.g. the userspace parts of the host VMM.
+
+* _Privileged guest software adversary_: This includes guest software components
+  executing in VS mode like e.g. the guest kernel.
+
+* _Privileged guest software adversary_: This includes guest software components
```
Comment (rsahita) [inline specification/security_model.adoc:80]:
```suggestion
* _Unprivileged guest software adversary_: This includes guest software components
```

Comment (sameo) [reply specification/security_model.adoc]: Fixed the copy-paste mistake, sorry.

File: specification/security_model.adoc (hunk 14 above)
Comment (sameo) [reply specification/security_model.adoc:7]: Right. That's kind of what I was trying to describe later in that section but your prose is definitely clearer. I used it, thanks.

File: specification/security_model.adoc [hunk 16]
```diff
@@ -1,9 +1,323 @@
...
+Each TDI must be exclusively assigned to no more than one TVM. However, a single
+TVM can simultaneously be assigned several TDIs, irrespective of whether they
+originate from the same physical device or not.
+
+As a TVM TCB does not include PCIe switches and bridges, a selective PCIe IDE
+stream must be setup to guarantee end-to-end confidentiality and integrity
+protection between a TVM and a TDI. The host VMM is responsible for setting up a
+single selective IDE stream between the host platform and any physical device
+from which one or more TDIs are assigned to TVMs. All TDIs within that physical
+device share the single PCIe IDE stream.
```
Comment (rsahita) [inline specification/security_model.adoc:46]: Should clarify that the host VMM is not in the TCB for the selective IDE stream - the endpoints for the selective IDE stream keys are configured by the TCB elements (TSM, RoT).

File: specification/security_model.adoc [hunk 17]
```diff
@@ -1,9 +1,323 @@
...
+  executing in VU mode like e.g. the guest application workload.
+
+* _Device firmware adversary_: This includes any firmware driving a device
+  within the system's PCIe topology. As PCIe transaction generators those device
+  firmware components can gain direct access to a TVM confidential memory.
+
+* _Hardware adversary_: This includes adversaries that have unlimited physical
+  access to the system and its devices, and can use simple or advanced hardware
+  attacks such as bus interposers, voltage/clock glitchers, microscopes, power
+  usage analyzer, etc. in order to access or tamper with the protected assets.
```
Comment (rsahita) [inline specification/security_model.adoc:92]: maybe better to split HW adversary into simple HW adversary and advanced HW adversary

File: specification/security_model.adoc [hunk 18]
```diff
@@ -1,9 +1,323 @@
...
+| In scope
+| Host component reads TVM confidential data
+
+5+| **Description** - A privileged host software component programs a TDI that
+    is assigned to a TVM. By accessing the device MMIO space, the host component
+    can program direct memory access destination addresses to either its own
+    address space or unintended parts of the TVM address space.
+    Device generated data intended to be copied to the TVM confidential memory
+    is respectively accessed by the host component instead or redirected to
+    unintened parts of the TVM address space.
```
Comment (rsahita) [inline specification/security_model.adoc:114]:
```suggestion
    unintended parts of the TVM address space.
```

**Review by rsahita — APPROVED** [review]
LGTM

File: specification/security_model.adoc (hunk 16 above)
Comment (sameo) [reply specification/security_model.adoc:46]: Added, thanks.

File: specification/security_model.adoc (hunk 17 above)
Comment (sameo) [reply specification/security_model.adoc]: Done.

File: specification/security_model.adoc (hunk 18 above)
Comment (sameo) [reply specification/security_model.adoc]: Good catch, fixed now. Thanks.

**Review by jyao1 — APPROVED** [review]

File: specification/security_model.adoc [hunk 19]
```diff
@@ -1,9 +1,438 @@
...
+  * Untrusted MMIO access to the TDI may still be allowed by the VMM for
+    ordinary VMs (if and only if the TDI is in the TDISP unlocked state).
+
+By means of the CoVE-IO guest ABI, TVMs are required to explicitly accept
+TDIs into their TCBs. Accepting or rejecting a TDI is a TVM specific decision,
+based on TVM specific set of verification policies and criteria. A TVM accepting
+a TDI does not imply that other TVMs on the same host platform would accept it
+as well.
+
+Multi-function, CoVE-IO-compliant devices may expose multiple TDIs, assigned to
```
Comment (sameo) [inline specification/security_model.adoc:34]: Multi-hosted? Multi-interfaces? Multi-user?

File: specification/security_model.adoc [hunk 20]
```diff
@@ -1,9 +1,438 @@
...
+different TVMs. It is the DSM responsibility to guarantee isolation between all
+assigned TDIs, on a per-TVM basis.
+
+Each TDI must be exclusively assigned to no more than one TVM. However, a single
+TVM can simultaneously be assigned several TDIs, irrespective of whether they
+originate from the same physical device or not.
+
+As a TVM TCB does not include PCIe switches and bridges, a selective PCIe IDE
+stream must be setup to guarantee end-to-end confidentiality and integrity
+protection between a TVM and a TDI. The host VMM is responsible for setting up a
```
Comment (sameo) [inline specification/security_model.adoc:44]: Be more precise on the host VMM responsibilities. It does not configure the whole thing, only stream id and device IDE config space.

File: specification/security_model.adoc [hunk 21]
```diff
@@ -1,9 +1,438 @@
...
+in the TDI before the host software stack can fully reclaim an assigned TDI.
+
+== Threat Model
+
+=== Assets
+
+The CoVE-IO security model aims at protecting the following assets:
+
+1. The TVM confidential data, which includes its confidential main memory, code
+   and execution state.
```
Comment (lfiolhais) [inline specification/security_model.adoc:67]: Apologies for being super nitpicky but I'm having trouble understanding what execution state is in this context. Do you mean the architectural state? The execution state of the TVM also contains the micro-architectural state and any "peripheral" (memory controllers, bridges, arbitrators, etc.) state. Maintaining the confidentiality of the whole execution state seems a tall order...

File: specification/security_model.adoc [hunk 22]
```diff
@@ -1,9 +1,438 @@
...
+
+2. An assigned TDI trusted I/O space that is mapped into the TVM address space,
+   also known as a TDI trusted MMIO.
+
+==== Security Objectives
+
+The CoVE-IO security model objectives is to protect the above-described TVM
+assests *confidentiality* and *integrity* from components outside of the TCB.
```
Comment (lfiolhais) [inline specification/security_model.adoc:75]: typo in assets

File: specification/security_model.adoc [hunk 23]
```diff
@@ -1,9 +1,438 @@
...
+
+*Availability* of these assets is out of the CoVE-IO security model scope.
 
 === Adversary Model
```
Comment (lfiolhais) [inline specification/security_model.adoc:79]: Instead of redefining the adversary model, can't we use a citation to the RISC-V adversary models from the global security model document?

File: specification/security_model.adoc (hunk 21 above)
Comment (jyao1) [reply specification/security_model.adoc:67]: I believe the aim is correct.
The CPU micro-architectural state should be guaranteed by the CPU SOC design and TSM.
The trusted peripheral device state should be guaranteed by the DSM.
The untrusted peripheral device shall not have any confidential data.

Comment (sameo) [reply specification/security_model.adoc]: > Apologies for being super nitpicky but I'm having trouble understanding what execution state is in this context. Do you mean the architectural state? 

*Thanks* for being picky.

> The execution state of the TVM also contains the micro-architectural state and any "peripheral" (memory controllers, bridges, arbitrators, etc.) state. Maintaining the confidentiality of the whole execution state seems a tall order...

I tried to clarify that now. Please let me know if the execution state definition needs further clarification.

File: specification/security_model.adoc (hunk 22 above)
Comment (sameo) [reply specification/security_model.adoc]: Fixed, thanks.

File: specification/security_model.adoc (hunk 23 above)
Comment (sameo) [reply specification/security_model.adoc:80]: We're not really redefining it, but rather contextualizing it for CoVE-IO. Also, we're splitting the software adversary groups between guest and host as they can trigger different attack vectors. And finally, the "startup code adversary" includes both host and device firmware, and implies that device firmware is a platform startup time only adversary. In the CoVE-IO case, device firmware can attack the TVM assets at any point in the platform lifecycle time.
So, to summarize it, I think the RVI security model is accurate but too generic for something like CoVE-IO.

File: specification/security_model.adoc (hunk 19 above)
Comment (sameo) [reply specification/security_model.adoc]: Fixed now.

File: specification/security_model.adoc (hunk 20 above)
Comment (sameo) [reply specification/security_model.adoc]: Clarified.
